app/controllers/repositories_controller.rb

```diff
@@ -30,46 +30,60 @@
 require 'SVG/Graph/Bar'
 require 'SVG/Graph/BarHorizontal'
 require 'digest/sha1'
+require_dependency 'open_project/scm/adapters'
 
-class ChangesetNotFound < Exception; end
-class InvalidRevisionParam < Exception; end
+class ChangesetNotFound < Exception
+end
+class InvalidRevisionParam < Exception
+end
 
 class RepositoriesController < ApplicationController
   include PaginationHelper
 
   menu_item :repository
-  menu_item :settings, only: :edit
+  menu_item :settings, only: [:edit, :destroy_info]
   default_search_scope :changesets
 
-  before_filter :find_repository, except: :edit
-  before_filter :find_project, only: :edit
+  before_filter :find_project, only: [:create, :update, :edit]
+  before_filter :find_repository, except: [:edit, :update, :create, :destroy, :destroy_info]
+  before_filter :find_project_by_project_id, only: [:destroy, :destroy_info]
   before_filter :authorize
   accept_key_auth :revisions
 
-  rescue_from Redmine::Scm::Adapters::CommandFailed, with: :show_error_command_failed
+  rescue_from OpenProject::Scm::Exceptions::ScmError, with: :show_error_command_failed
 
   def edit
-    @repository = @project.repository
-    if !@repository
-      @repository = Repository.factory(params[:repository_scm])
-      @repository.project = @project if @repository
-    end
-    if request.post? && @repository
-      @repository.attributes = params[:repository]
-      @repository.save
+    service = Scm::RepositoryFactoryService.new(@project, params)
+    if service.build_temporary
+      @repository = service.repository
+    else
+      logger.error("Cannot create repository for #{params[:scm_vendor]}")
+      flash.now[:error] = service.build_error
     end
 
-    menu_reload_required = if @repository.persisted? && !@project.repository
-                             @project.reload # needed to reload association
-                           end
+    respond_to do |format|
+      format.js { render 'repositories/settings/repository_form' }
+    end
+  end
 
+  def update
+    @repository = @project.repository
+    update_repository(params.fetch(:repository, {}))
     respond_to do |format|
-      format.js do
-        render template: '/projects/settings/repository',
-               locals: { project: @project,
-                         reload_menu: menu_reload_required }
-      end
+      format.js { render 'repositories/settings/repository_form' }
+    end
+  end
+
+  def create
+    service = Scm::RepositoryFactoryService.new(@project, params)
+    if service.build_and_save
+      @repository = service.repository
+      flash[:notice] = l('repositories.create_successful')
+    else
+      flash[:error] = service.build_error
     end
+
+    render js: "window.location = '#{settings_repository_tab_path}'"
```

Comment:
This should be wrapped in a `respond_to` to profit from rails default responses (e.g. 406) for unsupported mime types.